### ISSUE TITLE
Can't Login (2.0.0 Beta 4)

### ISSUE THREAD

**Issue by msmcgillis** [fisharebest/webtrees#2731]
Trying to take a look at webtrees 2.0.0 Beta 4.

Setup:
* Fedora 30 Server running Docker.
* A container running current mariadb image.
* A container based on fedora 30 image with this Dockerfile:

```
FROM fedora:30
MAINTAINER matthew@mcgillis.org
RUN dnf -y install httpd php php-gd php-json php-xml php-mbstring php-pecl-mcrypt php-pecl-zip php-pdo php-mysqlnd php-intl; dnf -y clean all; rm -rf /var/www; mkdir /run/php-fpm; sed -i "s/post_max_size\s*=\s*.*/post_max_size=100M/" /etc/php.ini
EXPOSE 80
CMD /sbin/php-fpm;/usr/sbin/httpd -DNO_DETACH -DFORGROUND
```
* mount a copy of webtrees-2.0.0-beta.4.zip contents to /var/www/html

Connect to server begin configuration get to the page to add administrator enter all the information.

It then takes me to a login page and I enter the user/pass to sign in and just end up back at the login page.

When tracing through the code I can see that when I enter the user name and password to sign in the CheckCsrf middleware is failing because the page CSRF value is not matching a null Session CSRF value. 

I can see in the database that Session data does exist for the WT_SESSION value. Although one curious thing is in the cookie I have two WT_SESSION values.

For some reason, the SessionDatabaseHandler read never returns a value. If I remove the ->where clause of DB::table('session') I can get all the records back but including the where clause never returns anything.

I'm guessing this is some sort of version or module issue of some sort.

```
# rpm -q -a | grep php
php-fpm-7.3.10-1.fc30.x86_64
php-cli-7.3.10-1.fc30.x86_64
php-7.3.10-1.fc30.x86_64
php-intl-7.3.10-1.fc30.x86_64
php-mbstring-7.3.10-1.fc30.x86_64
php-pecl-zip-1.15.4-3.fc30.x86_64
php-json-7.3.10-1.fc30.x86_64
php-recode-7.3.10-1.fc30.x86_64
php-devel-7.3.10-1.fc30.x86_64
php-common-7.3.10-1.fc30.x86_64
php-pdo-7.3.10-1.fc30.x86_64
php-gd-7.3.10-1.fc30.x86_64
php-xml-7.3.10-1.fc30.x86_64
php-pecl-mcrypt-1.0.2-2.fc30.x86_64
php-mysqlnd-7.3.10-1.fc30.x86_64
# php -v
PHP 7.3.10 (cli) (built: Sep 24 2019 09:20:18) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.10, Copyright (c) 1998-2018 Zend Technologies
# php -m
[PHP Modules]
bz2
calendar
Core
ctype
curl
date
dom
exif
fileinfo
filter
ftp
gd
gettext
hash
iconv
intl
json
libxml
mbstring
mcrypt
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
readline
recode
Reflection
session
SimpleXML
sockets
SPL
sqlite3
standard
tokenizer
wddx
xml
xmlreader
xmlwriter
xsl
zip
zlib

[Zend Modules]
```

**Comment by fisharebest** [comment]
There have been a huge number of changes since beta.4.  beta.5 is fairly imminent.
Are you able to test with the latest snapshot?
https://github.com/fisharebest/webtrees/archive/master.zip

**Comment by msmcgillis** [comment]
master.zip has the same issue.

Using the previously outlined setup: dropped the webtrees database and recreated it. Stopped and deleted the webtrees container. Created a new directory with master.zip contents created new webtrees container with master.zip contents mounted /var/www/html. Started going through the setup and after administration information is entered end up at the login page. Continually get returned to the login page when I enter the user/password and click sign in.

**Comment by fisharebest** [comment]
I have very little experience with Docker.

This sounds like cookies not being set correctly.  Wrong path/domain/port?

During setup, the base URL is stored in `data/config.ini.php`.  Is this correct?

**Comment by msmcgillis** [comment]
In the original message, I indicated the curious thing was my browser for some reason had two WT_SESSION values.

As an experiment, I tried a different browser safari and it let me in.

So I went back to chrome and restarted it. After that, I now only see one WT_SESSION value and it works.

Not sure what caused the two WT_SESSION values but I know in the traces the SessionDatabaseHandler read call was getting called twice. The first call always had the wrong WT_SESSION  value and the second call had the correct one. Wonder if somehow the first call was invalidating the second.

Not sure how the browser ever got two values to begin with but things seem good now.

**Comment by fisharebest** [comment]
I had a similar problem during development recently.

I installed the site at `/foo`, and then removed it.
I then installed it at `/foo/bar`.

The old cookie from `/foo` still existed in the browser's cache, and was sent to the new site at `/foo/bar`, preventing logins from working. 

> but things seem good now.

Thanks for the update.

I'm not sure there's anything we can do about this.  It's a bit of an edge case, and will disappear as soon as the cookies time out.

**Comment by msmcgillis** [comment]
I agree it is somewhat of an edge case but if you wanted to deal with it you could in the CSRF check. Before you validate the CSRF see if the browser has multiple WT_SESSION values that can be detected in the COOKIE information. If you see multiple values you could toss up an error instructing the user to exit the browser and return because of gremlins or something :).

If you wanted to make it transparent you could also do that. Where ever we process the COOKIES you could again detect the duplicate WT_SESSION values do some testing against the database to make the best guess of which value is correct then update the COOKIE to only have one value and remove the bad one.

Certainly not a high priority but addressable if you like.

**Comment by fisharebest** [comment]
I plan to implement different session drivers (e.g. using [APC](https://www.php.net/apc)).
I'll take another look at this then.

**Comment by msmcgillis** [comment]
I'm guessing something in the current code is adding to WT_SESSION rather than always forcing it to a single value. It could be in the session driver or the cookie manager.

**Comment by fisharebest** [comment]
> multiple WT_SESSION values that can be detected in the COOKIE information

Testing with firefox, I find that if I set two cookies with the same name, only the first one is included in the HTTP request.

Even if it sent both cookies, PHP doesn't seem to have a mechanism for handling it.
`$_COOKIE` is a one-dimensional array of strings, indexed by cookie name.

If you can think of a solution to this, please open a new issue.

**Comment by kpine** [comment]
Adding a comment because I think this might affect more people who are using reverse proxies.

In 1.7.15 I used the `HTTP=On` environment override to allow my https proxy to work with non-https Webtrees. When I upgraded to 2.0.0, I removed the env and updated the `config.ini.php` to use `base_url`. I found this in the upgrade guide, and read various issues that describe this. It worked fine, except I could not login. All of the assets were loading but logging in would just redirect with no login.

I checked my server logs and I see Firefox sending two `WT_SESSION` cookies being sent. I presume only one of the cookies is being accepted. I opened a private window and was able to login; only one cookie was sent. Looking at [Session.php](https://github.com/fisharebest/webtrees/blob/b262b3d3bf1186c52400e33a30ce62ac9143cd43/app/Session.php#L58), it looks like sessions are generated using the URL path including `base_url`. I am theorizing that a 2nd session and cookie are generated because I switched to using `base_url`. The 1.7.15 cookie was probably generated in a way that didn't use my hostname? If other proxy user like myself are switching to `base_url`, they will probably run into this problem.

The workaround is to clear the site cookies. Because no solution was proposed here, would it be  helpful to at least include a tip to clear cookies in the upgrade notes?
